mingw-w64-git/git-wrapper.c

```diff
@@ -41,6 +42,199 @@ static void print_error(LPCWSTR prefix, DWORD error_number)
+    if (wlen >= 4 && wstr[0] == L'\\' && wstr[1] == L'?' && wstr[2] == L'?' &&
+        wstr[3] == L'\\') {
+      /* Starts with \??\ */
+      if (wlen >= 6 &&
+          ((wstr[4] >= L'A' && wstr[4] <= L'Z') ||
+           (wstr[4] >= L'a' && wstr[4] <= L'z')) &&
+          wstr[5] == L':' && (wlen == 6 || wstr[6] == L'\\')) {
+        /* \??\<drive>:\ */
+        wstr += 4;
+        wlen -= 4;
+
+      } else if (wlen >= 8 && (wstr[4] == L'U' || wstr[4] == L'u') &&
+                 (wstr[5] == L'N' || wstr[5] == L'n') &&
+                 (wstr[6] == L'C' || wstr[6] == L'c') && wstr[7] == L'\\') {
+        /* \??\UNC\<server>\<share>\ - make sure the final path looks like */
+        /* \\<server>\<share>\ */
+        wstr += 6;
+        wstr[0] = L'\\';
+        wlen -= 6;
+      }
+    }
+    if (wlen >= maxsize) {
+      break;
+    }
+    wcsncpy(buf, wstr, wlen);
+    ret = TRUE;
+  } break;
+  case IO_REPARSE_TAG_MOUNT_POINT: {
+    wstr =
+        respbuf->MountPointReparseBuffer.PathBuffer +
+        (respbuf->MountPointReparseBuffer.SubstituteNameOffset / sizeof(WCHAR));
+    wlen =
+        respbuf->MountPointReparseBuffer.SubstituteNameLength / sizeof(WCHAR);
+    /* Only treat junctions that look like \??\<drive>:\ as symlink. */
+    /* Junctions can also be used as mount points, like \??\Volume{<guid>}, */
+    /* but that's confusing for programs since they wouldn't be able to */
+    /* actually understand such a path when returned by uv_readlink(). */
+    /* UNC paths are never valid for junctions so we don't care about them. */
+    if (!(wlen >= 6 && wstr[0] == L'\\' && wstr[1] == L'?' && wstr[2] == L'?' &&
+          wstr[3] == L'\\' &&
+          ((wstr[4] >= L'A' && wstr[4] <= L'Z') ||
+           (wstr[4] >= L'a' && wstr[4] <= L'z')) &&
+          wstr[5] == L':' && (wlen == 6 || wstr[6] == L'\\'))) {
+      SetLastError(ERROR_SYMLINK_NOT_SUPPORTED);
+      break;
+    }
+
+    /* Remove leading \??\ */
+    wstr += 4;
+    wlen -= 4;
+    if (wlen >= maxsize) {
+      break;
+    }
+    wcsncpy(buf, wstr, wlen);
+    ret = TRUE;
+  } break;
+  case IO_REPARSE_TAG_APPEXECLINK: {
+    if (respbuf->AppExecLinkReparseBuffer.StringCount != 0) {
+      LPWSTR szString = NULL, szTarget = NULL;
+      szString = (LPWSTR)respbuf->AppExecLinkReparseBuffer.StringList;
+      ULONG i = 0;
+      for (; i < respbuf->AppExecLinkReparseBuffer.StringCount; i++) {
+        if (i == 2) {
+          szTarget = szString;
+        }
+        szString += wcslen(szString) + 1;
+      }
+      if (wlen >= maxsize || szTarget == NULL) {
+        break;
+      }
+      wcsncpy(buf, szTarget, wlen);
+      ret = TRUE;
+    }
+  } break;
+  default:
+    break;
+  }
+  free(respbuf);
+  return ret;
+}
+
+static void find_exe_realpath(LPWSTR exepath, int count) {
+  WCHAR mdexe[MAX_PATH];
+  /* get the installation location */
+  GetModuleFileName(NULL, mdexe, MAX_PATH);
+  if (win_readlink(mdexe, exepath, count)) {
```

Comment:
I wonder whether we can take a shortcut and avoid a lot of code by using `GetFinalPathByHandleW()` here... According to https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getfinalpathnamebyhandlew it follows symbolic links...